packages/plugin-data/src/api/data/tree/exportApi.ts

```diff
@@ -0,0 +1,607 @@
+        }
+        return key;
+      });
+    });
+
+    // pre-load object metadata
+    const promises = Object.keys(this.objectTypeRegistry).map((key) => this.loadMetadata(key));
+    return Promise.all(promises).then(() => recordList);
+  }
+
+  private async processRecordList(recordList: any, parentRef?: any): Promise<any> {
+    // holds transformed sobject tree
+    const sobjectTree = { records: [] };
+
+    // visit each record in the list
+    const processRecordsFn = (record: any) => (): Promise<any> => this.processRecords(parentRef, record, sobjectTree);
+    const recordFns = recordList.records.map((record: any) => processRecordsFn(record));
+
+    // TODO: do we really need sequentialExecute?  Can't we just use Promise.all()?
+    await sequentialExecute(recordFns);
+    this.logger.debug(JSON.stringify(sobjectTree, null, 4));
+    return sobjectTree;
+  }
+
+  private async processRecords(parentRef: any, record: any, sobjectTree: any): Promise<any> {
+    // incremented every time we visit another record
+    const objRefId = this.incrementTypeRefIndex(record.attributes.type);
+
+    // add the attributes for this record, setting the type and reference
+    const treeRecord: any = {
+      attributes: {
+        type: record.attributes.type,
+        referenceId: objRefId,
+      },
+    };
+
+    // store the reference in a map with the record id
+    this.saveRecordRef(record, objRefId);
+
+    // handle each record attribute
+    await this.processRecordAttributes(record, treeRecord, objRefId);
+
+    if (parentRef && this.config.plan) {
+      const parentFieldName = getString(parentRef, 'fieldName', '');
+      if (!treeRecord[parentFieldName]) {
+        treeRecord[parentFieldName] = parentRef.id;
+      }
+    }
+    // add record to tree
+    sobjectTree.records.push(treeRecord);
+
+    return sobjectTree;
+  }
+
+  // Generate object type reference (<ObjectType>Ref<Counter>)
+  private incrementTypeRefIndex(type: string): string {
+    if (!this.typeRefIndexes[type]) {
+      this.typeRefIndexes[type] = 0;
+    }
+
+    return `${type}Ref${++this.typeRefIndexes[type]}`;
+  }
+
+  private async processRecordAttributes(record: any, treeRecord: any, objRefId: string): Promise<any> {
+    const promises = Object.keys(record).map((key) => this.processRecordAttribute(record, key, treeRecord, objRefId));
+
+    return Promise.all(promises).then(() => treeRecord);
+  }
+
+  private async processRecordAttribute(record: any, key: string, treeRecord: any, objRefId: string): Promise<any> {
+    return Promise.resolve().then(() => {
+      const field = record[key];
+
+      // skip attributes and id.  Data import does not accept records with IDs.
+      if (key === 'attributes' || key === 'Id') {
+        // If this is an attributes section then we need to add an object reference
+        this.saveRecordRef(record, objRefId);
+      } else {
+        return this.loadMetadata(record.attributes.type)
+          .then((metadata) => {
```

Comment:
Anyway we can refactor this to get rid of the promise chain?

Reply:
Yes but I'd like to save that for the refactor work item since it's not super straight forward.